lib/client/mfa.go

```diff
@@ -0,0 +1,144 @@
+// If promptDevicePrefix is set, it will be printed in prompts before "security
+// key" or "device". This is used to emphasize between different kinds of
+// devices, like registered vs new.
+func PromptMFAChallenge(ctx context.Context, proxyAddr string, c *proto.MFAAuthenticateChallenge, promptDevicePrefix string) (*proto.MFAAuthenticateResponse, error) {
+	if promptDevicePrefix != "" {
+		promptDevicePrefix = strings.TrimSpace(promptDevicePrefix) + " "
+	}
+
+	switch {
+	// No challenge.
+	case c.TOTP == nil && len(c.U2F) == 0:
+		return &proto.MFAAuthenticateResponse{}, nil
+	// TOTP only.
+	case c.TOTP != nil && len(c.U2F) == 0:
+		totpCode, err := prompt.Input(os.Stdout, os.Stdin, fmt.Sprintf("Enter an OTP code from a %sdevice", promptDevicePrefix))
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		return &proto.MFAAuthenticateResponse{Response: &proto.MFAAuthenticateResponse_TOTP{
+			TOTP: &proto.TOTPResponse{Code: totpCode},
+		}}, nil
+	// U2F only.
+	case c.TOTP == nil && len(c.U2F) > 0:
+		fmt.Printf("Tap any %ssecurity key\n", promptDevicePrefix)
+
+		return promptU2FChallenges(ctx, proxyAddr, c.U2F)
+	// Both TOTP and U2F.
+	case c.TOTP != nil && len(c.U2F) > 0:
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		type response struct {
+			kind string
+			resp *proto.MFAAuthenticateResponse
+			err  error
+		}
+		resCh := make(chan response)
+
+		go func() {
+			resp, err := promptU2FChallenges(ctx, proxyAddr, c.U2F)
+			select {
+			case resCh <- response{kind: "U2F", resp: resp, err: err}:
+			case <-ctx.Done():
+			}
+		}()
+
+		go func() {
+			totpCode, err := prompt.Input(os.Stdout, os.Stdin, fmt.Sprintf("Tap any %ssecurity key or enter a code from an %sOTP device", promptDevicePrefix, promptDevicePrefix))
+			res := response{kind: "TOTP", err: err}
+			if err == nil {
+				res.resp = &proto.MFAAuthenticateResponse{Response: &proto.MFAAuthenticateResponse_TOTP{
+					TOTP: &proto.TOTPResponse{Code: totpCode},
+				}}
+			}
+
+			select {
+			case resCh <- res:
+			case <-ctx.Done():
+			}
+		}()
+
+		for i := 0; i < 2; i++ {
+			select {
+			case res := <-resCh:
+				if res.err != nil {
+					log.WithError(res.err).Debugf("%s authentication failed", res.kind)
+					continue
+				}
+
+				// Print a newline after the TOTP prompt, so that any future
+				// output doesn't print on the prompt line.
+				fmt.Println()
+
+				return res.resp, nil
+			case <-ctx.Done():
+				return nil, trace.Wrap(ctx.Err())
+			}
+		}
+		return nil, trace.Errorf("failed to authenticate using all U2F and TOTP devices, rerun the command with '-d' to see error details for each device")
+	default:
+		return nil, trace.BadParameter("bug: non-exhaustive switch in promptMFAChallenge")
+	}
+}
+
+func promptU2FChallenges(ctx context.Context, proxyAddr string, challenges []*proto.U2FChallenge) (*proto.MFAAuthenticateResponse, error) {
+	facet := proxyAddr
+	if !strings.HasPrefix(proxyAddr, "https://") {
+		facet = "https://" + facet
+	}
```

Comment:
Is there a situation where we could be doing U2F logins in insecure mode i.e. HTTP? I can potentially see it for testing purposes, and adding `https` automatically when it isn't present would stop it from working, right?

Reply:
I don't think we support running a proxy without TLS.